test-data/unsigned.yaml

```diff
@@ -0,0 +1,973 @@
+code:
+  <start>: |
+    assume w1 > 1
+post: ["r1.type=number", "r1.value=[2, 3]"]
+---
+test-case: "assume 1 > 0 nop"
+pre: ["r1.type=number", "r1.value=1"]
+code:
+  <start>: |
+    assume r1 > 0
+post: ["r1.type=number", "r1.value=1"]
+---
+test-case: "assume 1 w> 0 nop"
+pre: ["r1.type=number", "r1.value=1"]
+code:
+  <start>: |
+    assume w1 > 0
+post: ["r1.type=number", "r1.value=1"]
+---
+test-case: "assume [1, INT_MAX] > 0 nop"
+pre: ["r1.type=number", "r1.value=[1, 9223372036854775807]"]
+code:
+  <start>: |
+    assume r1 > 0
+post: ["r1.type=number", "r1.value=[1, 9223372036854775807]"]
+---
+test-case: "assume [1, INT32_MAX] w> 0 nop"
+pre: ["r1.type=number", "r1.value=[1, 2147483646]"]
+code:
+  <start>: |
+    assume w1 > 0
+post: ["r1.type=number", "r1.value=[1, 2147483646]"]
+---
+test-case: "assume [0, INT_MAX] > 1 narrows to [2, INT_MAX]"
+pre: ["r1.type=number", "r1.value=[0, 9223372036854775807]"]
+code:
+  <start>: |
+    assume r1 > 1
+post: ["r1.type=number", "r1.value=[2, 9223372036854775807]"]
+---
+test-case: "assume [0, INT32_MAX] w> 1 narrows to [2, INT32_MAX]"
+pre: ["r1.type=number", "r1.value=[0, 2147483647]"]
+code:
+  <start>: |
+    assume w1 > 1
+post: ["r1.type=number", "r1.value=[2, 2147483647]"]
+---
+test-case: "assume [0, INT_MAX] > INT_MAX implies bottom"
+pre: ["r1.type=number", "r1.value=[0, 9223372036854775807]"]
+code:
+  <start>: |
+    r2 = 9223372036854775807ll
+    assume r1 > r2
+post: []
+messages:
+  - "1: Code is unreachable after 1"
+---
+test-case: "assume [0, INT32_MAX] w> INT32_MAX implies bottom"
+pre: ["r1.type=number", "r1.value=[0, 2147483647]"]
+code:
+  <start>: |
+    assume w1 > 2147483647
+post: []
+messages:
+  - "0: Code is unreachable after 0"
+---
+test-case: "assume [0, INT_MAX-4] >= INT_MAX-4 narrows to INT_MAX-4"
+pre: ["r1.type=number", "r1.value=[0, 9223372036854775803]"]
+code:
+  <start>: |
+    r2 = 9223372036854775803ll
+    assume r1 >= r2
+post: ["r1.type=number", "r1.value=9223372036854775803",
+        "r2.type=number", "r2.value=9223372036854775803",
+        "r2.value-r1.value<=0"]  # missing: r1.value-r2.value<=0; probably avoiding overflow in SplitDBM
+---
+test-case: "assume [0, INT32_MAX] w>= INT32_MAX narrows to INT32_MAX"
+pre: ["r1.type=number", "r1.value=[0, 2147483647]"]
+code:
+  <start>: |
+    r2 = 2147483647
+    assume w1 >= r2
+post: ["r1.type=number", "r1.value=2147483647",
+        "r2.type=number", "r2.value=2147483647",
+        "r2.value-r1.value<=0"]  # missing: r1.value-r2.value<=0; probably avoiding overflow in SplitDBM
+---
+test-case: "assume [-1, 1] > 0 nop"
+pre: ["r1.type=number", "r1.value=[-1, 1]"]
+code:
+  <start>: |
+    assume r1 > 0
+post: ["r1.type=number", "r1.value=[-1, 1]"]
+---
+test-case: "assume [-1, 1] w> 0 nop"
+pre: ["r1.type=number", "r1.value=[-1, 1]"]
+code:
+  <start>: |
+    assume w1 > 0
+post: ["r1.type=number", "r1.value=[-1, 1]"]
+---
+test-case: "assume [-1, 1] > 1 nop"
```

Comment:
why doesn't this test case narrow to -1?
Since [0, 1] > 1 is bottom (per test case at 570), that whole side of the range should be narrowed out.

Reply:
In a better implementation, yes it should narrow. Such an implementation may take all the different cases that are currently checked, use meet on each condition and then join all the result - exactly your logic here.

However, I thought it may be more error prone so I'd better get the simpler approach right first.
Also it may (or may not) be somewhat expensive, if we do the join and meet on the entire invariant.

Reply:
Note that such case split should be on exhaustive if we want to be sound, and mutually exclusive if we want to be precise. And still I suspect that in many cases the information will be lost after the join. Consider the case of [-1,2] > 1. The result must be [-1] | [2] which gives [-1,2] again.

Reply:
@caballa may have deeper insights (or corrections).

Reply:
I think it's fine to merge a less precise implementation now and maybe improve in future.

But I would treat ranges like [-X,Y] as a join between [-X,-1] and [0,Y] since -1 and 0 are quite different in terms of unsigned comparison.   So I would think that [-1, 1] > 1 should be much easier to handle than other arbitrary cases.

So logically splitting between -1 and 0:
[-1,1] > 1 => ((-1 > 1) join ([0,1] > 1)) => ((-1 > 1) join _|_) => (-1 > 1) => -1

Reply:
It only handles [-X,Y]>Y+k. Not [-X,Y+k]>Y. Maybe we can just check for the former explicitly. It might be just a small refinement of an existing check.

Reply:
Correct, that's what I'm suggesting.

Reply:
And from my understanding [-X,Y+k]>Y can't be handled in a single-interval domain.  Handling that would need a set-of-intervals domain.

Reply:
Yes it requires some non-convex domain.
(Y=0 is a long-standing motivation we have :)